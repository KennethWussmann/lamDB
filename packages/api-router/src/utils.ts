import { isExecutableDefinitionNode, OperationDefinitionNode, OperationTypeNode, parse } from 'graphql';
import { Request, Response, sha1Hash } from '@lamdb/core';
import { Request as ExpressRequest, Response as ExpressResponse } from 'express';

export const graphQlErrorResponse = (message: string): Response => ({
  status: 400,
  body: JSON.stringify({
    data: null,
    errors: [
      {
        message,
      },
    ],
  }),
  headers: {
    'content-type': 'application/json',
  },
});

export const fromExpressRequest = (expressRequest: ExpressRequest): Request => ({
  method: expressRequest.method,
  body: expressRequest.body,
  headers: Object.fromEntries(
    Object.entries(expressRequest.headers).map(([key, value]) => [key, Array.isArray(value) ? value[0] : value ?? '']),
  ),
  path: expressRequest.path,
});

export const applyToExpressResponse = (response: Response, expressResponse: ExpressResponse) => {
  Object.entries(response.headers).forEach(([key, value]) => expressResponse.setHeader(key, value));
  expressResponse.status(response.status);
  expressResponse.send(response.body);
};

export const getOperationInfo = (
  request: Request,
): { name: string; type: OperationTypeNode; hash: string | undefined } | undefined => {
  if (!request.body) {
    return undefined;
  }

  let parsedDocument;
  try {
    parsedDocument = JSON.parse(request.body);
  } catch {
    return undefined;
  }

  const document = parse(parsedDocument.query, {
    noLocation: true,
  });

  const executableNodes = document.definitions.filter(isExecutableDefinitionNode);

  if (executableNodes.length === 0) {
    throw new Error('Document does not include executable nodes');
  }

  const operationNodes: OperationDefinitionNode[] = executableNodes
    .filter((node: any) => !!node?.operation)
    .map((node) => node as OperationDefinitionNode);

  if (operationNodes.length === 0) {
    throw new Error('Document does not include operation nodes');
  }
  if (operationNodes.length > 1) {
    throw new Error('Document includes multiple operation nodes: Only one is supported');
  }
  return {
    name: parsedDocument.operationName ?? operationNodes[0].name ?? '',
    type: operationNodes[0].operation,
    hash: request.body ? sha1Hash(request.body) : undefined,
  };
};
