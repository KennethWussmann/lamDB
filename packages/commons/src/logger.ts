// istanbul ignore file
import { createLogger as createWinstonLogger, format } from 'winston';
import { Console } from 'winston/lib/winston/transports';

export type LogLevel = 'trace' | 'debug' | 'info' | 'warn' | 'error' | 'fatal';

export const createLogger = (
  meta: Record<string, unknown> | undefined = undefined,
  logLevel: LogLevel = (process.env.LOG_LEVEL as LogLevel) ?? 'info',
) =>
  createWinstonLogger({
    level: logLevel,
    format: format.combine(format.timestamp(), format.json()),
    defaultMeta: meta,
    transports: [new Console()],
  });
