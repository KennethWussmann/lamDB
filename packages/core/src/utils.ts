import { createHash } from 'crypto';
import { access } from 'fs/promises';

export const sha1Hash = (operation: string): string => {
  const hashsum = createHash('sha1');
  hashsum.update(operation);
  return hashsum.digest('hex').toString();
};

export const exists = async (file: string) => {
  try {
    await access(file);
    return true;
  } catch {
    return false;
  }
};

export const errorLog = (e: unknown): object => {
  const error = e as Error;
  return {
    error: {
      name: error.name,
      message: error.message,
      cause: error.cause,
      stack: error.stack,
    },
  };
};

export const getDatabaseUrl = (databaseFilePath: string) => `file:${databaseFilePath}?pool_timeout=5`;
