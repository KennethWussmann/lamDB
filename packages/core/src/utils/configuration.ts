// istanbul ignore file
import { config } from 'dotenv';

config();
export class LamDBConfiguration {
  get migrationEngineBinaryPath() {
    return assertEnv('MIGRATION_ENGINE_BINARY_PATH');
  }
  get migrationEngineForceMigration() {
    return assertEnv('MIGRATION_ENGINE_FORCE_MIGRATION', 'false').toLowerCase() === 'true';
  }
  get queryEngineLibraryPath() {
    return assertEnv('QUERY_ENGINE_LIBRARY_PATH');
  }
  get databasePath() {
    return assertEnv('DATABASE_PATH');
  }
  get prismaSchemaPath() {
    return assertEnv('PRISMA_SCHEMA_PATH');
  }
  get operationOptimization() {
    return assertEnv('OPERATION_OPTIMIZATION', 'false').toLowerCase() === 'true';
  }

  get server() {
    return new LamDBServerConfiguration();
  }
}

export class LamDBServerConfiguration {
  get allowCors() {
    return assertEnv('ALLOW_CORS', 'false').toLowerCase() === 'true';
  }
  get proxyPort() {
    return parseInt(assertEnv('PROXY_PORT', '4000'), 10);
  }
}

const assertEnv = (name: string, defaultValue: string | undefined = undefined): string => {
  const value = process.env[name] ?? defaultValue;
  if (!value) {
    throw new Error(`Environment variable ${name} not set, but required`);
  }
  return value;
};
