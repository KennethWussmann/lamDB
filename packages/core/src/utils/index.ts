export * from './utils';
export * from './configuration';
