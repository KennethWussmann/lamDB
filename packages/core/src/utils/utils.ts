export const getDatabaseUrl = (databaseFilePath: string) => `file:${databaseFilePath}?pool_timeout=3&connection_limit=1`;
