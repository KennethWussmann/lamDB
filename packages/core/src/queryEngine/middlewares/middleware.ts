import { Logger } from 'winston';
import { Request, Response } from '../../utils';
import { QueryEngine } from '../queryEngine';

export type MiddlewareContext = {
  request: Request;
  queryEngine: QueryEngine;
  logger: Logger;
};
export type MiddlewareNextFunction = (context: MiddlewareContext) => Promise<Response>;
export type QueryEngineProxyMiddleware = (
  context: MiddlewareContext,
  next: MiddlewareNextFunction,
) => Promise<Response> | Response;

export const executeMiddlewares = async (
  context: MiddlewareContext,
  middlewares: QueryEngineProxyMiddleware[],
  execute: (context: MiddlewareContext) => Promise<Response>,
): Promise<Response> => {
  const exec = async (context: MiddlewareContext, middlewares: QueryEngineProxyMiddleware[]): Promise<Response> => {
    if (!middlewares.length) {
      return await execute(context);
    }
    const middleware = middlewares[0];
    const response = await middleware(context, async (ctx) => await exec(ctx, middlewares.slice(1)));
    return response;
  };

  if (middlewares.length === 0) {
    return await execute(context);
  }

  return await exec(context, middlewares);
};

export interface Middleware {
  handle: QueryEngineProxyMiddleware;
}
