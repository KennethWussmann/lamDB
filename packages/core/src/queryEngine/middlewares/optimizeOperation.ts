import { optimizeDocuments as relayOperationOptimizer } from '@graphql-tools/relay-operation-optimizer';
import {
  ArgumentNode,
  astFromValue,
  DocumentNode,
  GraphQLInputType,
  GraphQLSchema,
  GraphQLType,
  Kind,
  OperationDefinitionNode,
  parse,
  print,
  typeFromAST,
  VariableDefinitionNode,
  visit,
} from 'graphql';
import { Logger } from 'winston';
import { errorLog, sha1Hash, Response, captureMethod, captureMethodSync, tracer } from '@lamdb/commons';
import { Middleware, MiddlewareContext, MiddlewareNextFunction } from './middleware';

class OptimizeOperationMiddleware implements Middleware {
  private inlineVariables = (schema: GraphQLSchema, variables: Record<string, unknown>, document: DocumentNode) =>
    captureMethodSync({
      segmentName: '### OptimizeOperationMiddleware.inlineVariables',
      method: () => {
        if (Object.keys(variables).length === 0) {
          return document;
        }
        const variableDefinitions: Record<string, GraphQLType> = {};
        return visit(document, {
          VariableDefinition: (node: VariableDefinitionNode) => {
            const type = typeFromAST(schema, node.type);
            if (type) {
              variableDefinitions[node.variable.name.value] = type;
            }
            return null;
          },
          Argument: (node: ArgumentNode) => {
            if (node.value.kind === Kind.VARIABLE) {
              if (!variables[node.value.name.value]) {
                // if supplied variables does not contain this variable, remove immediately
                return null;
              }
              const type = astFromValue(
                variables[node.value.name.value],
                variableDefinitions[node.value.name.value] as GraphQLInputType,
              );
              if (type?.kind === Kind.NULL) {
                // query engine does not like when nullable arguments actually are written out with null.
                // in case it's a variable and it's value is null, remove argument entirely.
                return null;
              }
              return { ...node, value: type };
            }
            return node;
          },
        });
      },
    });

  private generateOperationName = (document: DocumentNode) =>
    captureMethodSync({
      segmentName: '### OptimizeOperationMiddleware.generateOperationName',
      method: () =>
        visit(document, {
          OperationDefinition: (node: OperationDefinitionNode) => {
            if (!node.name) {
              // Prisma client does not supply operation names, but the relay optimizer requires one
              const operationName = `${node.operation}_${sha1Hash(print(node)).substring(0, 8)}`;
              tracer.getSegment().addMetadata('operationName', operationName);
              return {
                ...node,
                name: {
                  kind: Kind.NAME,
                  // Operation names have to start with letters, so we prepend query, mutation, subscription
                  value: operationName,
                },
              };
            }
            return node;
          },
        }),
    });

  private inlineFragments = (schema: GraphQLSchema, document: DocumentNode): DocumentNode =>
    captureMethodSync({
      segmentName: '### OptimizeOperationMiddleware.inlineFragments',
      method: () =>
        relayOperationOptimizer(schema, [document], {
          includeFragments: false,
          assumeValid: true,
          noLocation: true,
        })[0],
    });

  private optimizeDocument = (
    logger: Logger,
    schema: GraphQLSchema,
    variables: Record<string, unknown>,
    document: DocumentNode,
  ): DocumentNode =>
    captureMethodSync({
      segmentName: '### OptimizeOperationMiddleware.optimizeDocument',
      method: () => {
        try {
          const operationWithName = this.generateOperationName(document);
          const operationWithoutFragments = this.inlineFragments(schema, operationWithName);
          return this.inlineVariables(schema, variables, operationWithoutFragments);
        } catch (e) {
          logger.error('Failed to optimize operation. Returning unoptimized operation instead.', {
            document,
            ...errorLog(e),
          });
          tracer.addErrorAsMetadata(e as Error);
        }
        return document;
      },
    });

  private optimize = (logger: Logger, schema: GraphQLSchema, document: string): string =>
    captureMethodSync({
      segmentName: '### OptimizeOperationMiddleware.optimize',
      method: () => {
        const parsedDocument = JSON.parse(document);
        const optimizedOperation = this.optimizeDocument(
          logger,
          schema,
          parsedDocument?.variables ?? {},
          parse(parsedDocument.query, {
            noLocation: true,
          }),
        );
        try {
          return JSON.stringify({
            operationName: parsedDocument.operationName,
            variables: {},
            // TODO: This still throws an error
            // invalid AST Node: {}
            query: print(optimizedOperation),
          });
        } catch (e) {
          logger.error('Failed to print operation. Returning unoptimized operation instead.', {
            document,
            optimizedOperation,
            ...errorLog(e),
          });
          tracer.addErrorAsMetadata(e as Error);
        }
        return document;
      },
    });

  handle = async (context: MiddlewareContext, next: MiddlewareNextFunction): Promise<Response> =>
    await captureMethod({
      segmentName: '### OptimizeOperationMiddleware.handle',
      method: async () => {
        if (context.request.method.toLowerCase() !== 'post' || !context.request.body) {
          // skip middleware if not graphql or body empty
          return next(context);
        }
        return next({
          ...context,
          request: {
            ...context.request,
            body: this.optimize(context.logger, await context.queryEngine.getSchema(), context.request.body),
          },
        });
      },
    });
}

const middlewareClass = new OptimizeOperationMiddleware();
export const optimizeOperation = middlewareClass.handle.bind(middlewareClass);
