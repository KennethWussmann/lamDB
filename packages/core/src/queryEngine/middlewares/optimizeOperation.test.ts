import { createLogger } from '../../logger';
import { optimizeOperation } from './optimizeOperation';
import { getTestQueryEngine } from '../../../test/testQueryEngine';
import { operationToRequest } from '../../../test/utils';
import { MiddlewareContext } from './middleware';

let context: Pick<MiddlewareContext, 'queryEngine' | 'logger'>;

const findArticlesQuery = `
query findArticles(
  $where: ArticleWhereInput
  $orderBy: [ArticleOrderByWithRelationInput]
  $cursor: ArticleWhereUniqueInput
  $take: Int
  $skip: Int
  $distinct: [ArticleScalarFieldEnum]
) {
  findManyArticle(
    where: $where
    orderBy: $orderBy
    cursor: $cursor
    take: $take
    skip: $skip
    distinct: $distinct
  ) {
    id
    url
    title
    subtitle
    publication
    claps
  }
}`;
const findArticleByIdQuery = `
query findArticle(
  $id: Int!
) {
  findUniqueArticle(
    where: { id: $id }
  ) {
    id
    url
    title
    subtitle
    publication
    claps
  }
}`;

describe('optimizeOperation middleware', () => {
  beforeEach(() => {
    context = {
      queryEngine: getTestQueryEngine(),
      logger: createLogger({ name: 'test' }),
    };
  });

  it.each([
    [findArticlesQuery, { where: { title: { contains: 'Something' } } }],
    [findArticlesQuery, { where: { title: { contains: 'Something' } }, order: null, cursor: null }],
    [findArticlesQuery, { where: { title: { contains: 'Something' } }, order: null, skip: 1 }],
    // TODO: Variables are only partially supported. You can only pass entire input vars
    [findArticleByIdQuery, { id: 1 }],
  ])('inlines operation variables and removes variable definitions', async (query, variables) => {
    const next = jest.fn();

    await optimizeOperation(
      {
        ...context,
        request: operationToRequest(query, variables),
      },
      next,
    );

    expect(next).toHaveBeenCalled();

    const modifiedRequest = (next.mock.calls[0][0] as MiddlewareContext).request;
    expect(modifiedRequest).toMatchSnapshot();
  });

  it('inlines fragment and removes fragment definition', async () => {
    const next = jest.fn();

    await optimizeOperation(
      {
        ...context,
        request: operationToRequest(
          `
          fragment ArticleFragment on Article {
            id
            url
            title
            subtitle
            publication
            claps
          }
          query findAll {
            findManyArticle {
              ...ArticleFragment
            }
          }
          `,
        ),
      },
      next,
    );

    expect(next).toHaveBeenCalled();

    const modifiedRequest = (next.mock.calls[0][0] as MiddlewareContext).request;
    expect(modifiedRequest).toMatchSnapshot();
  });
});
