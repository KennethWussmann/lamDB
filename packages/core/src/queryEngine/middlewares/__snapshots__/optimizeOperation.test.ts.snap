// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optimizeOperation middleware inlines fragment and removes fragment definition 1`] = `
{
  "body": "{"operationName":"TestOperation","variables":{},"query":"query findAll {\\n  findManyArticle {\\n    id\\n    url\\n    title\\n    subtitle\\n    publication\\n    claps\\n    responses\\n  }\\n}"}",
  "method": "POST",
}
`;

exports[`optimizeOperation middleware inlines operation variables and removes variable definitions 1`] = `
{
  "body": "{"operationName":"TestOperation","variables":{},"query":"query findArticles {\\n  findManyArticle(where: {title: {contains: \\"Something\\"}}) {\\n    id\\n    url\\n    title\\n    subtitle\\n    publication\\n    claps\\n    responses\\n  }\\n}"}",
  "method": "POST",
}
`;

exports[`optimizeOperation middleware inlines operation variables and removes variable definitions 2`] = `
{
  "body": "{"operationName":"TestOperation","variables":{},"query":"query findArticles {\\n  findManyArticle(where: {title: {contains: \\"Something\\"}}) {\\n    id\\n    url\\n    title\\n    subtitle\\n    publication\\n    claps\\n    responses\\n  }\\n}"}",
  "method": "POST",
}
`;

exports[`optimizeOperation middleware inlines operation variables and removes variable definitions 3`] = `
{
  "body": "{"operationName":"TestOperation","variables":{},"query":"query findArticles {\\n  findManyArticle(where: {title: {contains: \\"Something\\"}}, skip: 1) {\\n    id\\n    url\\n    title\\n    subtitle\\n    publication\\n    claps\\n    responses\\n  }\\n}"}",
  "method": "POST",
}
`;
